### PR TITLE
JWT auth support

### LINKED ISSUE #136: JWT auth
Hello,

Jira also supports JWT autorization: https://developer.atlassian.com/cloud/jira/platform/authentication-for-apps/

is there plan to support this auth?

Currently I'm attaching authorization header in dirty way and I was able to perform simple GET request (getIssue method).
```js
  const jwt = require('atlassian-jwt')
  const makeRequest = Jira.prototype.makeRequest

  Jira.prototype.makeRequest = function (options, callback, successString) {
    const now = moment().utc()
    const url = new URL(options.uri)

    const tokenData = {
      'iss': 'connect-app', // app key
      'iat': now.unix(), // the time the token is generated
      'exp': now.add(3, 'minutes').unix(), // token expiry time (recommend 3 minutes after issuing)
      'qsh': jwt.createQueryStringHash({
        method: options.method,
        pathname: url.pathname
      })
    }

    const token = jwt.encode(tokenData, secret)

    if (!options.headers) {
      options.headers = {}
    }

    options.headers['Authorization'] = 'JWT ' + token

    return makeRequest.apply(this, [options, callback, successString])
  }
```

### PR DESCRIPTION
**Pull request by ItamarM**
fixes #136

### CODE REVIEW

**Comment by MrRefactoring** [comment]
It's awesome! Could you just do the indented PR separately?

**Comment by ItamarM** [comment]
@MrRefactoring  i switch the tabs back to 2 - sorry had default turned on here :)

**Comment by ItamarM** [comment]
@MrRefactoring sorry, i guess one of my plugins over-written this. How about now?

File: index.js [hunk 1]
```diff
@@ -219,6 +224,25 @@ var JiraClient = module.exports = function (config) {
                 pass: config.basic_auth.password
             };
         }
+    } else if (config.jwt) {
+      if (config.jwt.secret && config.jwt.iss) {
+        let expiry_time_seconds = 3 * 60; // expires in 3 minutes on default
```
Comment (MrRefactoring) [inline index.js]: You can write `const expiry_time_seconds = config.expiry_time_seconds || 3 * 60`

Comment (ItamarM) [reply index.js]: @MrRefactoring  you got it

**Comment by MrRefactoring** [comment]
@mochja can you test it?

**Comment by ItamarM** [comment]
@MrRefactoring if it helps i tested it internally. As long as you:
1. Supply the right configuration
1. Ask for at the right scopes (READ, WRITE if you intend to call write methods and so on).
1. Call the supported API paths (some paths are not supported by connected apps as documented by Atlassian - you will get 403 if you make such attempt).

File: types/config.d.ts [hunk 2]
```diff
@@ -21,6 +21,11 @@ export interface Config {
     token: string;
     token_secret: string;
   };
+  jwt?: {
+    iss?: string;
```
Comment (MrRefactoring) [inline types/config.d.ts]: Property should be required (`iss: string`)

Comment (ItamarM) [reply types/config.d.ts]: @MrRefactoring fixed

File: index.js [hunk 3]
```diff
@@ -141,16 +143,19 @@ var worklog = require('./api/worklog');
...
- * @param {string} [config.basic_auth.api_token] The api token of the user that will be authenticated. MUST be included 
+ * @param {string} [config.basic_auth.api_token] The api token of the user that will be authenticated. MUST be included
  *     if using email and api_token authentication.
  * @param {string} [config.oauth.consumer_key] The consumer key used in the Jira Application Link for oauth
  *     authentication.  MUST be included if using OAuth.
  * @param {string} [config.oauth.private_key] The private key used for OAuth security. MUST be included if using OAuth.
  * @param {string} [config.oauth.token] The VERIFIED token used to connect to the Jira API.  MUST be included if using
  *     OAuth.
  * @param {string} [config.oauth.token_secret] The secret for the above token.  MUST be included if using Oauth.
+ * @param {string} [config.jwt.iss] The Jira app key (can be found in the app descriptor). MUST be included
```
Comment (MrRefactoring) [inline index.js]: Also should be required

Comment (ItamarM) [reply index.js]: I used the same terminology as you used in other 'required' config keys. ("MUST be included").

Comment (MrRefactoring) [reply index.js]: We can remove the square brackets, this will mean that these fields are required

Comment (ItamarM) [reply index.js]: sure. i don't mind, but you used it on other keys. shell i do it?

Comment (MrRefactoring) [reply index.js]: Maybe we can describe it like this:

```
* @param {Object} [config.jwt]
* @param {string} config.jwt.iss
* @param {string} config.jwt.secret
```

Comment (ItamarM) [reply index.js]: @MrRefactoring fixed

**Comment by ItamarM** [comment]
@MrRefactoring - are you waiting for @mochja approval on this one?

**Comment by MrRefactoring** [comment]
I think we can try to add this feature

**Pull request merged**